### PR TITLE
🚀 Field parameterization

### PR DESCRIPTION
**Pull request by awtkns**
Screenshot 2023-11-21 at 2 33 12 PM

Can also make this optional

### CODE REVIEW

File: bananalyzer/runner/runner.py [hunk 1]
```diff
@@ -52,52 +70,91 @@ def _generate_name(self, example: Example) -> str:
...
+
+@pytest.fixture(scope="session")
+def event_loop():
+    loop = asyncio.new_event_loop()
+    yield loop
+    loop.close()
+
+
+@pytest.fixture(scope="session")
+def agent():
```
Comment (awtkns) [inline bananalyzer/runner/runner.py:117]: we are now using pytest fixtures instead of hacking with globals

File: bananalyzer/runner/runner.py [hunk 2]
```diff
@@ -22,25 +23,42 @@ class BananalyzerTest(BaseModel):
...
+    def generate_test(self, example: Example) -> BananalyzerTest:
         return BananalyzerTest(
             code=f"""
 @pytest.mark.asyncio
-async def {self._generate_name(example)}() -> None:
-    async with async_playwright() as p:
-        browser = await p.chromium.launch(headless={headless})
-        context = await browser.new_context()
-        example = get_example_by_url("{example.url}")
+class {self._generate_class_name(example)}:
```
Comment (awtkns) [inline bananalyzer/runner/runner.py:32]: Tests are now group as one Class per example. This allows us to run the agent once and then then run a series of tests against the result

File: bananalyzer/runner/null_agent_wrapper.py [hunk 3]
```diff
@@ -1,4 +1,5 @@
 import asyncio
+import random
```
Comment (asim-shrestha) [inline bananalyzer/runner/null_agent_wrapper.py:2]:
```suggestion

```

File: bananalyzer/__main__.py [hunk 4]
```diff
@@ -206,12 +221,10 @@ def main() -> int:
...
     # Load the desired tests
     generator = TestGenerator()
-    tests = [
-        generator.generate_test(example, args.headless) for example in filtered_examples
-    ]
+    tests = [generator.generate_test(example) for example in filtered_examples]
 
     # Run the tests
-    return run_tests(tests, agent, args.pytest_args)
+    return run_tests(tests, agent, args.pytest_args, args.headless, args.turbo)
```
Comment (asim-shrestha) [inline bananalyzer/__main__.py:227]: should package new params in pytest_args

File: bananalyzer/runner/runner.py [hunk 5]
```diff
@@ -22,25 +20,42 @@ class BananalyzerTest(BaseModel):
...
             example=example,
         )
 
+    def _generate_eval_test(self, eval_: Eval, i: int) -> str:
+        if eval_.type == "json_match" and isinstance(eval_.expected, dict):
+            return f"""
+    @pytest.mark.parametrize("key", {list(eval_.expected.keys())})
+    async def test_match_field(self, key, result) -> None:
+        assert self.example.evals[{i}].expected[key] == result[key]
+
```
Comment (asim-shrestha) [inline bananalyzer/runner/runner.py:51]: nit: Ideally we can encapsulate this stuff with the eval somewhere. Will need to change two places now whenever evals are added/deleted

File: bananalyzer/__main__.py [hunk 6]
```diff
@@ -87,6 +88,18 @@ def parse_args() -> Args:
...
+        "--quiet",
+        action="store_true",
+        help="Will decrease the verbosity of pytest",
+    )
+    parser.add_argument(
+        "--turbo",
+        action="store_true",
+        help="Will run tests in a single browser instance. This is faster but less "
+        "reliable",
+    )
```
Comment (asim-shrestha) [inline bananalyzer/__main__.py:102]: What does less reliable mean? What are the failure modes? Why would I want to use this if it's not reliable?

Also, can we call this `--single_browser_context` or something more obvious? 

```suggestion
    parser.add_argument(
        "--single_browser_instance",
        action="store_true",
        help="Run tests in a single browser instance as opposed to creating a browser instance per test. This is faster but less reliable as test contexts will occasionally bleed into each other, causing tests to fail",
    )
```

File: bananalyzer/__main__.py [hunk 7]
```diff
@@ -87,6 +88,18 @@ def parse_args() -> Args:
         default=[],
         help="A list of ids to skip tests on, separated by commas",
     )
+    parser.add_argument(
+        "-q",
+        "--quiet",
+        action="store_true",
+        help="Will decrease the verbosity of pytest",
```
Comment (asim-shrestha) [inline bananalyzer/__main__.py:95]:
```suggestion
        help="Will decrease the verbosity of pytest. By default we run with the `--v` pytest param.",
```

File: bananalyzer/runner/runner.py (hunk 5 above)
Comment (awtkns) [reply bananalyzer/runner/runner.py:51]: so you think evals should be responsible for code gen too?

File: bananalyzer/__main__.py (hunk 4 above)
Comment (awtkns) [reply bananalyzer/__main__.py:227]: these are not args that will be passed to pytest. dont think they should be combined imho

Comment (asim-shrestha) [reply bananalyzer/__main__.py:227]: oh true

**Review by asim-shrestha — APPROVED** [review]

File: bananalyzer/__main__.py (hunk 6 above)
Comment (awtkns) [reply bananalyzer/__main__.py:102]: done

File: bananalyzer/runner/runner.py [hunk 8]
```diff
@@ -22,25 +20,42 @@ class BananalyzerTest(BaseModel):
...
+"""
+        return f"""
+    async def test_{eval_.type}(self, result) -> None:
+        self.example.evals[{i}].eval_results(None, result)
+
+"""
+
     def _generate_name(self, example: Example) -> str:
```
[the remark this answers is not in the document]
Comment (awtkns) [reply bananalyzer/runner/runner.py]: done

File: bananalyzer/runner/null_agent_wrapper.py (hunk 3 above)
Comment (awtkns) [reply bananalyzer/runner/null_agent_wrapper.py]: done

**Pull request merged**